File: 543.diameter-of-binary-tree.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=543 lang=cpp
 *
 * [543] Diameter of Binary Tree
 */
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
private:
    int checkDepth(TreeNode *root, int &diameter)
    {
        if (!root)
        {
            return 0;
        }

        int left = checkDepth(root->left, diameter);
        int right = checkDepth(root->right, diameter);

        if (left + right > diameter)
            diameter = abs(left + right);
        return max(left, right) + 1;
    }

public:
    int diameterOfBinaryTree(TreeNode *root)
    {
        int diameter = 0;
        checkDepth(root, diameter);
        return diameter;
    }
};
// @lc code=end

File: Java/Grind75/Week 2/543.diameter-of-binary-tree.java
import java.util.*;

/*
 * @lc app=leetcode id=543 lang=java
 *
 * [543] Diameter of Binary Tree
 */
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    private int ans = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        armLength(root);
        return ans;
    }

    private int armLength(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int left = armLength(root.left);
        int right = armLength(root.right);
        ans = Math.max(left + right, ans);
        return Math.max(left, right) + 1;
    }
}
// @lc code=end
